examples/systemcalls/systemcalls.c

```diff
@@ -92,6 +162,83 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
  *   The rest of the behaviour is same as do_exec()
  *
 */
+    // Opening the output file
+    fd = open(outputfile, O_WRONLY | O_TRUNC | O_CREAT, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
+
+    // Error checking for opening file
+    if (fd < 0) 
+    { 
+        perror("Error opening file"); 
+        return false; 
+    }
+
+    // Forking the child and testing error condition
+    pid = fork();
+
+    if(pid == -1)
+    {
+        printf("Error occured while forking a process with errno %d\n", errno);
+        return false;
+    }
+    
+    // If child process, exec command
+    if(pid == 0)
+    {
+        int ret;
+
+        // Duplicating file descriptor and checking for error
+        if (dup2(fd, 1) < 0) 
+        { 
+            perror("Error with dup2"); 
+            close(fd);
+            exit(EXIT_FAILURE);
+        }
+
+
+        // Execute command and test error condition
+        ret = execv(command[0], command);
+
+        if(ret == -1)
+        {
+            perror("Eror executing the program with execv()");
+            exit(EXIT_FAILURE);
+        }
+    }
+    
+    // If parent process, wait for child to terminate
+    if(pid > 0)
+    {
+        //Close file descriptor
+        close(fd);
+
+        wait_pid = wait(&status);
+
+        // Error testing
+        if (wait_pid == -1)
+        {
+            perror ("Error with wait()");
+            return false;
+
+        }
+        else
+        {
+            // Check if child exited normally
+            if(WIFEXITED(status))
+            {
+                // Check if child exited with a return value 0
+                if(WEXITSTATUS(status) != 0)
+                {
+                    return false;
```

Comment:
Missing cleanup in these code paths.
Also see discussion in class about avoiding duplication between these two very similar functions.